library/psa_crypto_slot_management.c

```diff
@@ -58,13 +58,110 @@ MBEDTLS_STATIC_ASSERT(PSA_KEY_ID_VOLATILE_MAX < MBEDTLS_PSA_KEY_ID_BUILTIN_MIN |
 
 
 
+#if defined(MBEDTLS_PSA_KEY_STORE_DYNAMIC)
+
+/* Dynamic key store.
+ *
+ * The key store consists of multiple slices.
+ *
+ * The volatile keys are stored in variable-sized tables called slices.
+ * Slices are allocated on demand and deallocated when possible.
+ * The size of slices increases exponentially, so the average overhead
+ * (number of slots that are allocated but not used) is roughly
+ * proportional to the number of keys (with a factor that grows
+ * when the key store is fragmented).
+ *
+ * One slice is dedicated to the cache of persistent and built-in keys.
+ * For simplicity, they are separated from volatile keys. This cache
+ * slice has a fixed size and has the slice index KEY_SLOT_CACHE_SLICE_INDEX,
+ * located after the slices for volatile keys.
+ */
+
+/* Size of slice 0 containing the cache of persistent and built-in keys. */
+#define PERSISTENT_KEY_CACHE_COUNT MBEDTLS_PSA_KEY_SLOT_COUNT
+
+/* Volatile keys are stored in slices 1 through KEY_SLICE_COUNT inclusive.
+ * Each slice is twice the size of the previous slice.
+ * Volatile key identifiers encode the slice number as follows:
+ *     bits 30..31:  0b10 (mandated by the PSA Crypto specification).
+ *     bits 25..29:  slice index (0...KEY_SLOT_VOLATILE_SLICE_COUNT-1)
+ *     bits 0..24:   slot index in slice
+ */
+#define KEY_ID_SLOT_INDEX_WIDTH 25u
+#define KEY_ID_SLICE_INDEX_WIDTH 5u
+
+#define KEY_SLOT_VOLATILE_SLICE_BASE_LENGTH 16u
+#define KEY_SLOT_VOLATILE_SLICE_COUNT 22u
+#define KEY_SLICE_COUNT (KEY_SLOT_VOLATILE_SLICE_COUNT + 1u)
+#define KEY_SLOT_CACHE_SLICE_INDEX KEY_SLOT_VOLATILE_SLICE_COUNT
+
+#if KEY_ID_SLICE_INDEX_WIDTH + KEY_ID_SLOT_INDEX_WIDTH > 30
+#error "Not enough room in volatile key IDs for slice index and slot index"
+#endif
+#if KEY_SLICE_COUNT >= (1 << KEY_ID_SLICE_INDEX_WIDTH) - 1
```

Comment:
I think this is not quite correct for 2 reasons:

1. `1 << KEY_ID_SLICE_INDEX_WIDTH` does not itself fit into `KEY_ID_SLICE_INDEX_WIDTH`, so a `-1` is needed
2. `KEY_SLOT_VOLATILE_SLICE_COUNT` does not take into account the 1 slice we use for persistent keys, so we must use `KEY_SLICE_COUNT`

With these tweaks we would have:
```suggestion
#if KEY_SLICE_COUNT > (1 << KEY_ID_SLICE_INDEX_WIDTH) - 1
```
Which still gains one extra slice. It's possible that this is intended as an 'invalid slice' value, but I can see no reason for this to be necessary. The actual bit width limitation is only for the packed format within key IDs, so we would only need it if we want to have a key ID with an 'invalid slice' within it.

I'll change to my tweaked version of this to gain an extra slice.

Reply:
I disagree.
> 1. 1 << KEY_ID_SLICE_INDEX_WIDTH does not itself fit into KEY_ID_SLICE_INDEX_WIDTH, so a -1 is needed

This check should be equivalent to the following check below:
https://github.com/Mbed-TLS/mbedtls/blob/e8199f574c2a1360a11fa4d6c71d0f1b14a285df/library/psa_crypto_slot_management.c#L106-L108

The largest possible index is `COUNT - 1`, and the largest we can fit in `WIDTH` is `(1 << WIDTH) - 1`, so the correct check is `#if COUNT - 1 > (1 << WIDTH) - 1 #error`, which is equivlent to `#if COUNT > (1 << WIDTH) #error`

> 2. KEY_SLOT_VOLATILE_SLICE_COUNT does not take into account the 1 slice we use for persistent keys, so we must use KEY_SLICE_COUNT

Only volatile keys encode the slice/slot ids inside the key ids. [Persistent keys' ids are chosen by the application](https://arm-software.github.io/psa-api/crypto/1.1/api/keys/ids.html) (not Mbed TLS), and thus we only use them as opaque handles (only the fact that they are persistent keys is encoded in the key id).

Since the persistent key slice only containes persistent keys, we never need to encode its slice id in a volatile key.

So the largest value we need to fit into `KEY_ID_SLICE_INDEX_WIDTH` is `KEY_SLOT_VOLATILE_SLICE_COUNT  - 1`.

Reply:
Ah, you're right. My apologies, I will change that.